Add a separate git-gui formula for the Tcl/Tk git UIs

arichardson opened the pull request:
As a tcl-tk dependency for git is not considered acceptable, this change
moves the gitk and git-gui tools to a new formula and sets the NO_TCLTK
flag when building git.

Additionally, this change ensures that we no longer install a .app for
git-gui. Instead, we use the homebrew-installed tcl-tk wish binary to run
git-gui and gitk.

Without this change, all dialogs that are shown by git-gui are blank
windows and can only be closed by memorizing the location of the buttons.

This fixes #36390 and #39987

This is yet another alternative approach to #42788 and #42789.

- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

arichardson commented on Formula/git-gui.rb:
```diff
@@ -0,0 +1,50 @@
+  homepage "https://git-scm.com"
+  # Note: these should be kept in sync with git.rb
+  url "https://www.kernel.org/pub/software/scm/git/git-2.25.0.tar.xz"
+  sha256 "c060291a3ffb43d7c99f4aa5c4d37d3751cf6bca683e7344ea407ea504d9a8d0"
```
I am not sure if it's possible to use something like `Git.sha256` and `Git.url` to prevent these from become out of sync?

SeekingMeaning replied:
This compiles
```ruby
  homepage "#{Formula["git"].homepage}"
  url "#{Formula["git"].stable.url}"
  sha256 "#{Formula["git"].stable.checksum}"
  head "#{Formula["git"].head}", :shallow => false
```

arichardson replied:
Thanks, I've updated the PR.

jkoberg commented:
I need this fix! Updating brew has broken my development workflow, as `git gui` no longer launches.  Upset that @MikeMcQuaid stated by fiat that `git gui` is "useless" and used by "a minority of users" and closed/locked other bug reports - I'd like to see the measurements he bases that on!  It's like saying "a minority of users drive Subarus - so it's OK if our roads don't work with them"

SMillerDev commented:
You can very easily maintain any software you want in your own tap. Mike's comments only apply to the main tap.

MikeMcQuaid commented on Formula/git-gui.rb:
```diff
@@ -0,0 +1,49 @@
+    # See https://github.com/Homebrew/homebrew-core/issues/36390
+    tcl_bin = Formula["tcl-tk"].opt_bin
+    args += %W[
+      TKFRAMEWORK=/we/do/not/want/to/install/a/.app/for/git-gui
```
Could this be `/dev/null` instead?

arichardson replied:
Seems to work.

MikeMcQuaid commented on Formula/git-gui.rb:
```diff
@@ -0,0 +1,49 @@
+class GitGui < Formula
+  desc "Tcl/Tk UI for the git revision control system"
+  homepage "#{Formula["git"].homepage}"
```
Please set these all to non-variable strings.

MikeMcQuaid commented on Formula/git-gui.rb:
```diff
@@ -0,0 +1,49 @@
+    # We also tell git to use the homebrew-installed wish binary from tcl-tk.
+    # See https://github.com/Homebrew/homebrew-core/issues/36390
+    tcl_bin = Formula["tcl-tk"].opt_bin
+    args += %W[
```
Please make these part of the `args` array above as they are not e.g. dynamically assigned based on a conditional.

MikeMcQuaid commented on Formula/git-gui.rb:
```diff
@@ -0,0 +1,49 @@
+      TCLTK_PATH=#{tcl_bin}/wish
+    ]
+    system "make", "-C", "git-gui", "install", *args
+    system "make", "-C", "gitk-git", "install", *args
```
Do these install a single binary (each) or more than that? If a single binary maybe it's just worth doing `bin.install` instead?

arichardson replied:
They install multiple files each:
```
/usr/local/Cellar/git-gui/2.25.0
/usr/local/Cellar/git-gui/2.25.0/bin
/usr/local/Cellar/git-gui/2.25.0/bin/gitk
/usr/local/Cellar/git-gui/2.25.0/libexec
/usr/local/Cellar/git-gui/2.25.0/libexec/git-core
/usr/local/Cellar/git-gui/2.25.0/libexec/git-core/git-gui
/usr/local/Cellar/git-gui/2.25.0/libexec/git-core/git-citool
/usr/local/Cellar/git-gui/2.25.0/libexec/git-core/git-gui--askpass
/usr/local/Cellar/git-gui/2.25.0/share
/usr/local/Cellar/git-gui/2.25.0/share/git-gui
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/choose_font.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/branch_delete.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/remote_add.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/status_bar.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/browser.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/line.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/about.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/spellcheck.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/git-gui.ico
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/chord.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/win32.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/sshkey.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/mergetool.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/error.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/win32_shortcut.js
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/console.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/logo.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/date.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/transport.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/branch.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/merge.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/branch_checkout.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/encoding.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/branch_rename.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/shortcut.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/commit.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/remote.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/search.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/tclIndex
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/el.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/sv.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/ja.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/pt_br.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/ru.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/fr.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/pt_pt.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/vi.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/nb.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/bg.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/hu.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/zh_cn.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/de.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/msgs/it.msg
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/tools.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/database.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/checkout_op.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/choose_repository.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/blame.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/option.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/diff.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/index.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/remote_branch_delete.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/branch_create.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/choose_rev.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/themed.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/tools_dlg.tcl
/usr/local/Cellar/git-gui/2.25.0/share/git-gui/lib/class.tcl
/usr/local/Cellar/git-gui/2.25.0/share/gitk
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/sv.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/ja.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/pt_br.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/ru.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/fr.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/pt_pt.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/vi.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/ca.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/bg.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/es.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/hu.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/zh_cn.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/de.msg
/usr/local/Cellar/git-gui/2.25.0/share/gitk/lib/msgs/it.msg
```

MikeMcQuaid commented on Formula/git-gui.rb:
```diff
@@ -0,0 +1,49 @@
+    # `git gui` from the command line.
+    local_git_core = libexec/"git-core"
+    local_git_core.each_child do |f|
+      mv f.to_s, bin
```
Use `bin.install` instead and hardcode the things that are being installed.

MikeMcQuaid commented on Formula/git-gui.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+  test do
+    system libexec/"git-core/git-gui", "--version"
```
Use `bin` here. Is there anything that can be done beyond just `--version` on the GUIs?

arichardson replied:
Not very much since they pop up the GUI  with any other option and I'm not sure how we would test that.

MikeMcQuaid commented on Formula/git.rb:
```diff
@@ -75,6 +76,11 @@ def install
+    # The git-gui and gitk tools are installed by a separate formula (git-gui)
+    # to avoid a dependency on tcl-tk and to avoid using the broken system
+    # tcl-tk (see https://github.com/Homebrew/homebrew-core/issues/36390)
+    args += %w[NO_TCLTK=1]
```
Add this to the `args` array above and note the `git-gui`/`gitk` stuff in the caveats.

MikeMcQuaid commented:
> Upset that @MikeMcQuaid stated by fiat that `git gui` is "useless"

This selective quote is incredibly misleading. What I actually said:

- "If it's completely useless we should remove it from the package"
- "That may be an argument for keeping it around if it isn't completely useless"

> I'd like to see the measurements he bases that on!

https://formulae.brew.sh/formula/git displays installation counts. Given `git-gui` is broken and we've had very few comments about this: the majority do not use `git-gui`.

> and closed/locked other bug reports

This is how this open source project is run. Feel free to run your open source projects however you choose and we'll do the same.
---

@jkoberg Your drive-by comments on your personal needs are neither helpful or welcome. If you (or anyone else) want to be helpful you can work on helping with PRs and asking what's required to get them merged rather than just jumping in and criticising people who spend their free time building stuff you use for free.

timblechmann commented:
> https://formulae.brew.sh/formula/git displays installation counts. Given `git-gui` is broken and we've had very few comments about this: the majority do not use `git-gui`.

fwiw, i haven't commented about the broken package as i was busy with other stuff, and switching to something like `git-cola` was a stop-gap workaround for me (i still prefer `git-gui` for various reasons, especially startup time, but also as `gitk` and `git-gui` are working on virtually every platform that git runs on).

that said: i'm happy to see that someone takes over this issue and is working on a fix.

arichardson commented:
I hope I've addressed all issues in this latest version.
I force-pushed the PR instead of adding a fixup commit but I'm not sure what the recommended homebrew approach is here.

MikeMcQuaid commented on Formula/git-gui.rb:
```diff
@@ -0,0 +1,39 @@
+
+  test do
+    system bin/"git-gui", "--version"
+    system bin/"gitk", "--version"
```
Removed this when I pushed.

MikeMcQuaid commented:
Thanks so much for your contribution! Without people like you submitting PRs we couldn't run this project. You rock, @arichardson!

bxt commented:
Awesome! I did a `brew install git-gui` and it made `git gui` work again for me 🎉🎉🎉 Thank you so much to everyone involved! 🙏

irq commented:
Thanks for fixing this! It works, but I get the following output in the console when I start `git-gui`.

```
2020-01-28 07:53:53.901 wish[52701:457151] CoreText note: Client requested name ".SFNSMono-Regular", it will get Times-Roman rather than the intended font. All system UI font access should be through proper APIs such as CTFontCreateUIFontForLanguage() or +[NSFont systemFontOfSize:].
2020-01-28 07:53:53.901 wish[52701:457151] CoreText note: Set a breakpoint on CTFontLogSystemFontNameRequest to debug.
2020-01-28 07:53:53.929 wish[52701:457151] CoreText note: Client requested name ".SF NS Mono", it will get Times-Roman rather than the intended font. All system UI font access should be through proper APIs such as CTFontCreateUIFontForLanguage() or +[NSFont systemFontOfSize:].
2020-01-28 07:53:54.021 wish[52701:457151] CoreText note: Client requested name ".SFNSMono-Regular", it will get Times-Roman rather than the intended font. All system UI font access should be through proper APIs such as CTFontCreateUIFontForLanguage() or +[NSFont systemFontOfSize:].
2020-01-28 07:53:54.021 wish[52701:457151] CoreText note: Client requested name ".SF NS Mono", it will get Times-Roman rather than the intended font. All system UI font access should be through proper APIs such as CTFontCreateUIFontForLanguage() or +[NSFont systemFontOfSize:].
2020-01-28 07:53:54.021 wish[52701:457151] CoreText note: Client requested name ".SFNSMono-Regular", it will get Times-Roman rather than the intended font. All system UI font access should be through proper APIs such as CTFontCreateUIFontForLanguage() or +[NSFont systemFontOfSize:].
2020-01-28 07:53:54.022 wish[52701:457151] CoreText note: Client requested name ".SFNSMono-Bold", it will get Times-Roman rather than the intended font. All system UI font access should be through proper APIs such as CTFontCreateUIFontForLanguage() or +[NSFont systemFontOfSize:].
2020-01-28 07:53:54.022 wish[52701:457151] CoreText note: Client requested name ".SF NS Mono", it will get Times-Roman rather than the intended font. All system UI font access should be through proper APIs such as CTFontCreateUIFontForLanguage() or +[NSFont systemFontOfSize:].
2020-01-28 07:53:54.022 wish[52701:457151] CoreText note: Client requested name ".SFNSMono-Regular", it will get Times-Roman rather than the intended font. All system UI font access should be through proper APIs such as CTFontCreateUIFontForLanguage() or +[NSFont systemFontOfSize:].
```

airdrummingfool commented:
Thanks for this fix!

@irq That output is a Wish (TCL/TK) issue. I've seen similar since before switching to this new `git-gui` brew package. To get around it, I created aliases in my global `.gitconfig` file that pipe all stderr output to `/dev/null`:

```
[alias]
        g = "!f() { git-gui $@ 2> /dev/null & }; f"
        k = "!f() { gitk $@ 2> /dev/null & }; f"
```

now I launch git gui with `git g` and gitk with `git k`. I imagine you could override the built-in `gui` alias, but I'm not sure.
